src/analyses/apron/apronPriv.apron.ml

```diff
@@ -43,6 +42,74 @@ module type S =
     val finalize: unit -> unit
   end
 
+module type Converter =
+  functor (ADGlobal: ApronDomain.S2) -> functor (Priv: S) -> S
+
+(** Takes [ADGlobal], an apron-domain-module to be used for global invariants, a privatizataion [Priv], and returns a privatization that
+    internally uses [ADGlobal] to keep global invariants, while complying to the privatzation interface fitting to [ADLocal].
+    This is achieved by internally converting back and forth between representations of the different domains. *)
+module Converter: Converter = functor (ADGlobal: ApronDomain.S2) -> functor (Priv: S) -> functor (ADLocal: ApronDomain.S2) ->
+struct
+  module Priv = Priv(ADGlobal)
+  module D = Priv.D
+  module G = Priv.G
+
+  type t = ADLocal.t
+  type global_act = (ADGlobal.t, D.t) ApronDomain.aproncomponents_t
+  type apron_components_t = (t, D.t) ApronDomain.aproncomponents_t
+
+  let to_local apr = ADLocal.of_other_dom (module ADGlobal.Man) apr
+  let to_global apr = ADGlobal.of_other_dom (module ADLocal.Man) apr
+
+  let to_local_components (comp: global_act): apron_components_t = { apr = to_local comp.apr; priv = comp.priv}
+  let to_global_components (comp: apron_components_t): global_act = { apr = to_global comp.apr; priv = comp.priv}
+
+  let name () = "Converter " ^ "(Privatization: " ^ (Priv.name ()) ^ ", local domain: " ^ (ADLocal.Man.name ()) ^ ", global domain: " ^ (ADGlobal.Man.name ()) ^ ")"
+
+  let startstate () = Priv.startstate ()
+
+  let should_join a b = Priv.should_join (to_global_components a) (to_global_components b)
+
+  let read_global ask getg st g x =
+    let r = Priv.read_global ask getg (to_global_components st) g x in
+    to_local r
+
+  let write_global ?(invariant=false) ask getg sideg st g x =
+    let r = Priv.write_global ask getg sideg (to_global_components st) g x in
+    to_local_components r
```

Comment:
If the local state `st` contains some non-interval constraints, then converting it to interval manager will lose those and converting from intervals back to octagons won't restore them. I think the same applies to all the other transfer functions too. Or am I misunderstanding this converter?

I think one way to make it right is to give each `Priv` module two `AD` arguments for the local (inside components) and globals (inside global invariants). And then have each of them do the conversions at appropriate places when they're transferring information from one to the other.

A simpler approach might be to lift `getg` and `sideg` with this converter and have them perform the conversion on the global parts, not the local parts. That is, `sideg` will take the octagon and convert it to a box and `getg` will do vice versa (create an octagon which just has box constraints).